### ISSUE TITLE
Allow easy status-go consumption from nodejs environment

### ISSUE THREAD

**Issue by jeluard** [status-im/ideas#2]
Idea: DEV#002
Title: Easy status-go consumption from nodejs environment
Status: Draft
Created: 2017-11-08

## Summary

A simple `nodejs` module encapsulating all ethereum interactions.

## Vision

Status accesses ethereum via a go module built on top of ethereum. This go module is then used through Java and C bridges.
As we are considering using `nodejs` based platforms (notably `electron` for `status-desktop`) it makes sense to offer a `nodejs` module.

**NOTE** Mobile support, while desirable, is explicitly out of scope. For this reason existing glue code in `status-go` will have to be kept for backward compatibility.

## Swarm Participants

* Lead contributor: @flexsurfer 
* Contributor: @adambabik   (__.5__)
* Evaluator: @yenda  

### Easy nodejs integration

As a `nodejs` app writer I want to integrate with `status-node-ethereum` by simply adding the proper dependency in my `package.json`.

`web3` (used by `status-react`, webview (for DApps) and inside jails) will rely on a `Web3 provider` offered by the node module. Other internal APIs can be provided differently depending on implementation convenience.

```javascript
import Web3 from 'web3.js';
import Statusgo from 'node-status';

const status = new StatusProvider({type: "native", id: "ropsten"});
const web3 = new Web3(status.attach());

web3.eth.doSomething();
```

### Easy testing / experimentation

A simple REPL could offer a great platform for quick experimentation and testing. This is similar to what `geth -console` does.
NodeJS offers ample libraries to facilitate this.

```javascript
./status-node-ethereum
web3.version;
> "1.0.0"
```

### Unified API (mid-term)

As a go developer I only have to manage a single interface for `status-go` consumers.

# Goals & Implementation Plan

`status-react` team will be the main consumer of this API so they should provide input on what is expected. `status-go` team is responsible from the technical implementation.

Security is our top priority. In particular no internal capacity is leaked to the existing JSON-RPC interface.
A strong focus on performance will be put. There should be little overhead in using this approach compared to the existing one (say .1ms extra). While mobile support is out of scope it is important to keep it in mind.

Particularly a in-process implementation is favored (everything is contained in the library).

### MVP

Let's constraint work on this to 1 week: we should either have a working implementation by then or very concrete steps on how to achieve this.

A simple module that can be loaded and demonstrates viability of the bindings technical solution.

```javascript
import Web3 from 'web3.js';
import Statusgo from 'node-status';

const status = new StatusProvider({type: "native", id: "ropsten"});
const web3 = new Web3(status.attach());

web3.version;
```

Goal Date: 2017-11-17
Description: a simple functional nodejs module

### Iteration #1 

`status-desktop` (built using electron) will be used as first consumer of this API. A simple version of chat should be built on top of `status-node-ethereum` to validate the viability of the approach.

Offer a [Web3 provider](https://github.com/ethereum/wiki/wiki/JavaScript-API#adding-web3)
Expose [management APIs](https://github.com/ethereum/go-ethereum/wiki/Management-APIs): `admin`, `debug`, `personal` to the extent they make sense.
Expose status API (https://github.com/status-im/status-go/blob/develop/lib/library.go)

Goal Date: 2017-11-24
Description: the final nodejs module

### Copyright
Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).

**Comment by jeluard** [comment]
I pledge focus 1 (20h/w) for 2 weeks.

**Comment by influx6** [comment]
I currently have no pending tasks on my list, so i can fully pledge (35h/w) for 2 weeks on this, while the rest (5h/w) is left for review work on other PRs.

**Comment by influx6** [comment]
Currently with my initial look which is still needing a bit of research, on the side of writing C++ binding for `status`, the questions that arise in my thoughts are:

1. Do we wish to expose to the other side and are we doing a `method-to-method` conversion, to allow the nodejs end able to reach into the go side.

2. Do we wish to expose a RPC like method which then replays commands from nodejs side to the go side, which might be more flexible for our needs and limits the amount of work to be exposed on the nodejs side. If this is possible, we get the benefit of implement custom objects for `eth`, `net`, `shh` that needs to interface directly with the go side.

Will research more on how extensions are equally written for node modules and compare my thoughts, your thoughts would be awesome to compare. :)

**Comment by adambabik** [comment]
I won't have time to work on this but I am happy to comment and share my thoughts :) 

As far as I know, we would need to compile status-go with cgo and use exported C bindings, just like we're doing it now for mobile. Glue code between status-go and node needs to be written in C++.

As far as I know, `web3.js` can be run in node environment. Having said that, I wouldn't try to expose web3 APIs directly from status-go but would try to use `web3.js` to connect to a running geth node. It will limit the scope of work to exposing functions like start/stop a geth node, create an account and so on.

**Comment by jeluard** [comment]
@influx6 great!

My first thoughts would be that we want to keep it flexible (probably more will be added) but perf is number 1 concern. We don't want to introduce a big overhead. 

If by RPC you mean something involving some form of serialization and network call it sounds too expensive.

**Comment by themue** [comment]
We're currently working for such a kind of API exposing using JSON-RPC, originally mostly intended for optimising and debugging. But it will allow to access the API.

@jeluard In which way is performance your largest concern? IMHO we don't talk about server optimisation with thousands of concurrent clients but about a simple 1-to-1 communication sending individual commands to a daemon in the background, don't we? Let me know if I'm wrong.

**Comment by jeluard** [comment]
@themue Context is all `web3` methods should be exposed so we don't know how they might be used by DApps. Status `chat` itself might grow to bigger usage over time. Also take into consideration that `status-desktop` might offer others usages we don't currently have (multiple tabs for instance).

I would say communication overhead should be in the low sub-ms order (like .2 ms).
Surely we can find some flexible solution that would offer the perf benefit of direct intra-process calls (here I don't foresee inter-process needs).

@themue Feel free to pledge some time if you are interested in the subject :)

**Comment by themue** [comment]
@jeluard Will benchmark the stuff, yes. My last project has been a server (so a larger number of concurrent users with a web UI) providing its services with a RESTful API serialising everything via JSON. Performance has been a smaller issue. So I'm optimistic. 😏

**Comment by influx6** [comment]
Am very inclined toward the doc style approach but if we choose to maintain c++ bindings then sure that a fare game as well.

**Comment by jeluard** [comment]
@influx6 What do you mean by doc style?

Let's make sure we investigate all possible options before jumping to conclusion. Quick googling reveals [node-go-require](https://github.com/sagiegurari/node-go-require) for instance.

**Comment by adambabik** [comment]
Worth to check but `node-go-require` looks like something to simply compile `.go` files into JavaScript code using [gopherjs](https://github.com/gopherjs/gopherjs). I doubt it can compile the whole status-go into JS :P

**Comment by jeluard** [comment]
Right we don't want that for sure ;) Didn't look in detail just was under the impression there is room for automating native bindings somehow.

**Comment by themue** [comment]
As mentioned above based on my integration experiences I'm alway for the most loose coupling possible. Here in this case like @influx6 said. So

```
[node.js]<-- JSON -->[statusd]
```

regardless if the protocol is JSON-RPC, is RESTful using own JSON documents, or is something proprietary we decide for. From my use cases for Go in backends (clouds and RESTful microservices with HTML/JS/CSS frontends) I don't expect performance troubles. If I'm right we're talking about one client / one statusd here, no huge scaling scenario. Correct me if I'm wrong here.

**Comment by jeluard** [comment]
@themue My only point is please let's list all available options with pros/cons then make a conscious decision.

**Comment by themue** [comment]
@jeluard Sure, absolutely on your side. Collecting options and arguments here, create a decision table, evaluate and compare these options, and the decide for a common agreed way.

**Comment by yenda** [comment]
I am in favor of extending the JSON-RPC protocol instead of working on native bindings, because it is more loosely coupled and woud make collaboration easier if we have a clearly defined API. I understand that if we have a node running it won't be hard to have websockets or http server (even reusing the existing one) for JSON-RPC, but how do we start/stop a node ? Also how is the initialization done (create directory for blockchain data for instance) ?

**Comment by divan** [comment]
So far the task seems to boil down to:
 1. implementing web3.js provider on JS side that will translate `web3.namespace.method` calls into `statusgo.CallRPC("namespace_method")`, similar to provider used in `status-react` - https://github.com/status-im/status-react/blob/8c044d9991223e75b1c7bc8549ed014b22047ca6/src/status_im/utils/web3_provider.cljs#L19
 2. Creating node.js-compatible module for status-go using native code support and Go's ability to compile into C shared library.

**Comment by jacqueswww** [comment]
I just saw that #7 could actually be picked as iteration 0 for this idea. If decided to go the route of using a JSON-RPC interface. Can these two ideas be done in parallel ?

**Comment by jeluard** [comment]
@jacqueswww Actually #7 has been extracted from this `idea` to speed up implementation and focus on the `nodejs` bindings. Surely it will follow soon after!

**Comment by jacqueswww** [comment]
@jeluard perfect! :smile_cat:

**Comment by oskarth** [comment]
This would be useful for running integration tests too on status-react side, as these are evaluated in NodeJS. For example, support for CreateAccount and similar would be useful, as opposed to using web3 JS API over HTTP-RPC only.

**Comment by jeluard** [comment]
@oskarth Yes the plan is to expose all those methods plus a bunch others. If you have some specific needs please share!

**Comment by yenda** [comment]
@jeluard the need to be to mimic the current Android and iOS bindings (or the other way around ) so they can be swapped with the nodejs ones for testing

**Comment by jeluard** [comment]
Shifting my focus away on other issues where I can provide more value.

**Comment by oskarth** [comment]
@flexsurfer @jeluard @yenda  @adambabik what's the state of this swarm? Is it still active? Is it done? On hold?

**Comment by flexsurfer** [comment]
we can hold it, we have a working version of https://github.com/status-im/status-nodejs lib, but we need to do some work later https://github.com/status-im/status-nodejs/issues

**Comment by flexsurfer** [comment]
let's close this idea, it was complete, we can continue work on status-nodejs lib in the scope of 12/DESKTOP idea
